### PR TITLE
Training Platform Core: minimal MLP end-to-end training on Siracusa

### PR DESCRIPTION
**Pull request by runwangdl**
Minimum viable **MLP end-to-end training** on Siracusa (untiled + tiled). Brings in only the framework + operators needed to build a complete forward → backward → gradient-accumulation → SGD graph on `simplemlp_train`. Targets `devel`.

**Deliberately excluded** (future follow-up PRs): GAP9 port, all Conv/Pool/Norm gradients, MSELoss, GroupNorm, PULPTrainlib submodule, FP32 forward kernels in `TargetLibraries/`.

## Added

- **Training codegen framework** under `DeeployTest/`
  - `testMVPTraining.py`, `testMVPOptimizer.py` — training / optimizer codegen entry points (parallel to the inference `testMVP.py`).
  - `generateTrainingNetwork.py`, `generateOptimizerNetwork.py` — untiled codegen entry points (parallel to `generateNetwork.py`).
  - `deeployTrainingRunner_siracusa.py`, `deeployTrainingRunner_tiled_siracusa.py` — end-to-end training runners, mirroring the inference `deeployRunner` pattern.
  - `testUtils/trainingUtils.py`, `testUtils/codeGenerateTraining.py`, `testUtils/deeployTrainingRunner.py` — training codegen helpers, shared-arena patching, L1 capacity checks, L3 hex-dump path, training subprocess orchestration.
- **Siracusa training harness** — `DeeployTest/Platforms/Siracusa/src/deeploytraintest.c` + CMake updates (top-level + Siracusa): `TRAINING`-gated build targets driving both TrainingNetwork and OptimizerNetwork across multiple steps with gradient accumulation, plus `-DTRAINING=ON`, `-DN_TRAIN_STEPS`, `-DN_ACCUM_STEPS`, `-DTRAINING_NUM_DATA_INPUTS` defines.
- **PULPOpen training operators**
  - `InPlaceAccumulatorV2` (ORT gradient-accumulation, optional lazy reset) — layer, parser, type checker, tile constraint, binding, template.
  - Canonical 2-output `SoftmaxCrossEntropyLoss` (loss scalar + log_prob tensor) — matching the ONNX Runtime training-graph signature. Includes parser, tile constraint, and template.
  - `SoftmaxCrossEntropyLossGrad` tile constraint.
  - `SGD` template update for L2 / L3 alias lowering.
- **Tiling support for in-place alias outputs** — `Targets/PULPOpen/Tiler.py` and `TilingExtension/TilerExtension.py` skip zero-sized in-place alias outputs from MiniMalloc (which rejects size-0 entries) and copy their `addrSpace` from the alias target after the solver runs.

## Changed

- `Deeploy/DeeployTypes.py` — `VariableBuffer.has_live_aliases` treats `is_input` / `is_output` buffers as live, so training input / output tensors survive the alias-liveness prune.
- `Deeploy/Targets/Generic/{Layers,Parsers,TypeCheckers}.py` — generic-layer side of the new `InPlaceAccumulatorV2` and canonical `SoftmaxCrossEntropyLoss` operators.
- `Deeploy/Targets/PULPOpen/{Bindings,Platform}.py` — register new training layer bindings and mapper lists; no inference-side changes.
- `DeeployTest/testUtils/core/{config,execution}.py` — +8 lines / 0 deletions total, all training hooks (L3 flash-image detection + `TRAINING` CMake flag propagation). Inference execution paths unchanged.

## Fixed

- `SoftmaxCrossEntropyLoss`** — an earlier design carried both a legacy 1-output form and a new 2-output form side by side, which tripped Deeploy's root-layer backtracking (it commits to the first parser whose `parseNode` succeeds and cannot retry sibling parsers if downstream `typeCheck` fails). Collapsed to a single canonical 2-output path and regenerated `Tests/Kernels/FP32/Softmax/CrossEntropy` ONNX + `outputs.npz` accordingly.

## Verification

```bash
# Untiled
python DeeployTest/deeployTrainingRunner_siracusa.py \
    -t /app/Onnx4Deeploy/onnx/model/simplemlp_train
# -> Errors: 0 out of 4   (all diff=0.000000)

# Tiled (L1=64000, L2=2000000)
python DeeployTest/deeployTrainingRunner_tiled_siracusa.py \
    -t /app/Onnx4Deeploy/onnx/model/simplemlp_train \
    --l1 64000 --l2 2000000
# -> Errors: 0 out of 4   (all diff=0.000000)
```

Reference losses: `loss 0: 2.123129`, `loss 1: 3.045749`, `loss 2: 2.914456`, `loss 3: 2.393848`.

## PR Merge Checklist

1. [x] The PR is rebased on the latest `devel` commit and pointing to `devel`.
2. [ ] Your PR reviewed and approved.
3. [x] All checks are passing.
4. [ ] The `CHANGELOG.md` file has been updated.
5. [x] If the docker was modified, change back its link after review.

### CODE REVIEW

File: Deeploy/Targets/Generic/Parsers.py [hunk 1]
```diff
@@ -2697,6 +2704,48 @@ def parseNodeCtxt(self,
...
+
+    Output:
+        0: output_buffer   - updated accumulation buffer (float tensor)
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def parseNode(self, node: gs.Node) -> bool:
+        # Require exactly 3 inputs (buffer, gradient, lazy_reset_grad) and 1 output
```
Comment (Victor-Jung) [inline Deeploy/Targets/Generic/Parsers.py]: NITPICK: Redundent comment

File: Deeploy/Targets/PULPOpen/Templates/FloatInPlaceAccumulatorV2Template.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from Deeploy.DeeployTypes import NetworkContext, NodeTemplate, OperatorRepresentation
+
+
+class _PULPInPlaceAccumulatorV2Template(NodeTemplate):
+    """True in-place InPlaceAccumulatorV2 template for PULP.
+
+    Writes the accumulation result into ``accum_buffer`` (the graph input).
+    ``data_out`` is registered as an alias of ``accum_buffer`` so the memory
+    allocator knows they share memory and will not free ``accum_buffer``
+    prematurely.
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/FloatInPlaceAccumulatorV2Template.py:16]: Is this behavior verified somewhere?

File: Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py [hunk 3]
```diff
@@ -2,9 +2,43 @@
...
+
+    def alignToContext(
+            self, ctxt: NetworkContext,
+            operatorRepresentation: OperatorRepresentation) -> Tuple[NetworkContext, OperatorRepresentation, List[str]]:
+        weight = ctxt.lookup(operatorRepresentation['weight'])
+        weight_updated = ctxt.lookup(operatorRepresentation['weight_updated'])
+
+        weight.aliases.add(weight_updated.name)
+        weight_updated.aliases.add(weight.name)
+        weight_updated._alias = weight.name
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py]: Why doing `weight_updated._alias = weight.name` explicitly? Lines 29 and 30 seem to do that already?

File: Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py [hunk 4]
```diff
@@ -2,9 +2,43 @@
...
+
+        # Make weight_updated share weight's allocation (no separate malloc),
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py]: This seems AI generated, even if the tensors were separated there would be no malloc as we play with memory arenas.

File: Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py [hunk 5]
```diff
@@ -2,9 +2,43 @@
...
+        # regardless of which memory level (L2 or L3) weight is placed in.
+        # The egress DMA then writes updated weights back to weight's address.
+        weight_updated.allocTemplate = NodeTemplate(" ${name} = (${type.typeName}) " + str(weight._instance) + ";")
+        weight_updated.deallocTemplate = NodeTemplate("")
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/SGDTemplate.py]: This looks like a big hack to me. Why is this necessary?

File: Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py [hunk 6]
```diff
@@ -4,27 +4,29 @@
 
 from Deeploy.DeeployTypes import NodeTemplate
 
+# Canonical SoftmaxCrossEntropyLoss: emits both a scalar mean loss and the
+# per-sample log_prob tensor.
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py]: NITPICK: Looks like AI-generated comments with no added value IMO.

File: Deeploy/Targets/PULPOpen/TileConstraints/InPlaceAccumulatorV2TileConstraint.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from Deeploy.CommonExtensions.DataTypes import uint16_t
+from Deeploy.DeeployTypes import NetworkContext, OperatorRepresentation
+from Deeploy.Targets.Generic.TileConstraints.BOPTileConstraint import BOPTileConstraint
+from Deeploy.TilingExtension.MemoryConstraints import NodeMemoryConstraint
+from Deeploy.TilingExtension.TilerModel import TilerModel
+from Deeploy.TilingExtension.TilingCodegen import AbsoluteHyperRectangle, HyperRectangle, TilingSchedule, \
+    VariableReplacementScheme
+
+
+class InPlaceAccumulatorV2TileConstraint(BOPTileConstraint):
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/TileConstraints/InPlaceAccumulatorV2TileConstraint.py:19]: I don't see a tile constraint that would require the tiles of the 2 inputs to match. How do you enforce that the input tiles are valid?

**Review by Victor-Jung — COMMENTED** [review]
First quick round of review. I have lots of questions already. I'll do a second round once these questions are answered.
